# Data Cleaning
It is the process of detecting and correcting corrupt or inaccurate records from a record-set, table, or database and refers to identifying incomplete, incorrect, inaccurate, or irrelevant parts of the data, and then replacing, modifying, or deleting the dirty or coarse data.
Data cleaning means fixing bad data in your data set.

Bad data could be:

* Empty cells
* Data in wrong format
* Wrong data
* Duplicates


Data Cleaning is a technique which helps to convert improper data into meaningful data. 
In short, Machine Learning is data-driven. With data cleaning in place, your Machine Learning model will perform better. So, it is important to process data before use. Without quality data, it is foolish to expect a correct output.

# Data Cleaning Techniques. 

* Remove Repeated Row 
* Missing value treatment 
* Removal of Irrelevant Data 
* Manual error while typing 
* Renaming Columns 

In [3]:
import pandas as pd
datafile="dirtydata.csv"
dataset = pd.read_csv(datafile)

In [4]:
dataset

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

In [3]:
#The data set contains some empty cells ("Date" in row 22, and "Calories" in row 18 and 28).
#The data set contains wrong format ("Date" in row 26).
#The data set contains wrong data ("Duration" in row 7).
#The data set contains duplicates (row 11 and 12).

In [4]:
dataset['Duration'].describe()

count     32.000000
mean      68.437500
std       70.039591
min       30.000000
25%       60.000000
50%       60.000000
75%       60.000000
max      450.000000
Name: Duration, dtype: float64

In [5]:
#For checking missing values, we have used the function is.null().
dataset.isnull()

Duration   Date  Pulse  Maxpulse  Calories
0      False  False  False     False     False
1      False  False  False     False     False
2      False  False  False     False     False
3      False  False  False     False     False
4      False  False  False     False     False
5      False  False  False     False     False
6      False  False  False     False     False
7      False  False  False     False     False
8      False  False  False     False     False
9      False  False  False     False     False
10     False  False  False     False     False
11     False  False  False     False     False
12     False  False  False     False     False
13     False  False  False     False     False
14     False  False  False     False     False
15     False  False  False     False     False
16     False  False  False     False     False
17     False  False  False     False     False
18     False  False  False     False      True
19     False  False  False     False     False
20     False  False  False     False     False
21     False  False  False     False     False
22     False   True  False     False     False
23     False  False  False     False     False
24     False  False  False     False     False
25     False  False  False     False     False
26     False  False  False     False     False
27     False  False  False     False     False
28     False  False  False     False      True
29     False  False  False     False     False
30     False  False  False     False     False
31     False  False  False     False     False

In [6]:
#One way to do this is to use a chained version the .isnull() method and the .sum() method:
print(dataset.isnull().sum())

Duration    0
Date        1
Pulse       0
Maxpulse    0
Calories    2
dtype: int64


# Empty Cells
Empty cells can potentially give you a wrong result when you analyze data.

In [7]:
# Return a new Data Frame with no empty cells:
df = dataset.dropna()
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

#By default, the dropna() method returns a new DataFrame, and will not change the original.
#If you want to change the original DataFrame, use the inplace = True argument:

In [24]:
#Example
#Remove all rows with NULL values:

# Replace Empty Values
Another way of dealing with empty cells is to insert a new value instead.
This way you do not have to delete entire rows just because of some empty cells.
The fillna() method allows us to replace empty cells with a value:

In [8]:
import pandas as pd
datafile="dirtydata.csv"
dataset = pd.read_csv(datafile)
dataset.fillna(130, inplace = True)
dataset

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112     130.0
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           130    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132     130.0
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

Replace Only For Specified Columns
The example above replaces all empty cells in the whole Data Frame.

To only replace empty values for one column, specify the column name for the DataFrame:

In [9]:
dataset["Calories"].fillna(130, inplace = True)


In [10]:
dataset

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112     130.0
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           130    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132     130.0
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

Replace Using Mean, Median, or Mode
A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

In [10]:
#Calculate the MEAN, and replace any empty values with it:

dataset = pd.read_csv('dirtydata.csv')

x = dataset["Calories"].mean()
print(x)

dataset["Calories"].fillna(x, inplace = True)

304.68


In [11]:
dataset

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'    100       120    300.00
18        45  '2020/12/18'     90       112    304.68
19        60  '2020/12/19'    103       123    323.00
20        45  '2020/12/20'     97       125    243.00
21        60  '2020/12/21'    108       131    364.20
22        45           NaN    100       119    282.00
23        60  '2020/12/23'    130       101    300.00
24        45  '2020/12/24'    105       132    246.00
25        60  '2020/12/25'    102       126    334.50
26        60      20201226    100       120    250.00
27        60  '2020/12/27'     92       118    241.00
28        60  '2020/12/28'    103       132    304.68
29        60  '2020/12/29'    100       132    280.00
30        60  '2020/12/30'    102       129    380.30
31        60  '2020/12/31'     92       115    243.00

# Data of Wrong Format
Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

To fix it, you have two options: remove the rows, or convert all cells in the columns into the same format.

# Convert Into a Correct Format
In our Data Frame, we have two cells with the wrong format. Check out row 22 and 26, the 'Date' column should be a string that represents a date:

Pandas has a to_datetime() method for this:



In [12]:
dataset = pd.read_csv('dirtydata.csv')

dataset['Date'] = pd.to_datetime(dataset['Date'])

dataset

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
22        45        NaT    100       119     282.0
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

#As you can see from the result, the date in row 26 was fixed, 
#but the empty date in row 22 got a NaT (Not a Time) value, 
#in other words an empty value. One way to deal with empty values is simply removing the entire row.
#The result from the converting in the example above gave us a NaT value, 
#which can be handled as a NULL value, 
#and we can remove the row by using the dropna() method.

In [13]:
dataset.dropna(subset=['Date'], inplace = True)
dataset


Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

# Wrong Data
"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

Sometimes you can spot wrong data by looking at the data set, because you have an expectation of what it should be.

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.



# Replacing Values
One way to fix wrong values is to replace them with something else.

In [14]:
dataset.loc[7, 'Duration'] = 45

In [15]:
dataset

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

For small data sets you might be able to replace the wrong data one by one, but not for big data sets.

To replace wrong data for larger data sets you can create some rules, e.g. set some boundaries for legal values, and replace any values that are outside of the boundaries.

In [15]:
#Loop through all values in the "Duration" column.
#If the value is higher than 120, set it to 120:

for x in dataset.index:
  if dataset.loc[x, "Duration"] > 60:
     dataset.loc[x, "Duration"] = 60
dataset        

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

In [16]:
#Delete rows where "Duration" is higher than 120:

for x in dataset.index:
  if dataset.loc[x, "Duration"] > 120:
    dataset.drop(x, inplace = True)

In [17]:
dataset

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

# Discovering Duplicates
Duplicate rows are rows that have been registered more than one time.

In [18]:
dataset

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

In [21]:
#row 11 and 12 are duplicates.

# To discover duplicates, we can use the duplicated() method.
# The duplicated() method returns a Boolean values for each row:
# Returns True for every row that is a duplicate, othwerwise False:

print(dataset.duplicated())

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


In [19]:
print(dataset.duplicated().sum())

1


# Removing Duplicates
To remove duplicates, use the drop_duplicates() method.

In [20]:
dataset.drop_duplicates(inplace = True)

In [23]:
dataset

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

In [23]:
# Dropping Based on a Subset of Columns
dataset = pd.read_csv('dirtydata.csv')
dataset = dataset.sort_values(by='Calories', ascending=False)
dataset

Duration          Date  Pulse  Maxpulse  Calories
1         60  '2020/12/02'    117       145     479.0
0         60  '2020/12/01'    110       130     409.1
4         45  '2020/12/05'    117       148     406.0
30        60  '2020/12/30'    102       129     380.3
14        60  '2020/12/14'    104       132     379.3
6         60  '2020/12/07'    110       136     374.0
21        60  '2020/12/21'    108       131     364.2
13        60  '2020/12/13'    106       128     345.3
2         60  '2020/12/03'    103       135     340.0
25        60  '2020/12/25'    102       126     334.5
10        60  '2020/12/11'    103       147     329.3
19        60  '2020/12/19'    103       123     323.0
23        60  '2020/12/23'    130       101     300.0
5         60  '2020/12/06'    102       127     300.0
17        60  '2020/12/17'    100       120     300.0
3         45  '2020/12/04'    109       175     282.4
22        45           NaN    100       119     282.0
29        60  '2020/12/29'    100       132     280.0
15        60  '2020/12/15'     98       123     275.0
9         60  '2020/12/10'     98       124     269.0
7        450  '2020/12/08'    104       134     253.3
12        60  '2020/12/12'    100       120     250.7
11        60  '2020/12/12'    100       120     250.7
26        60      20201226    100       120     250.0
24        45  '2020/12/24'    105       132     246.0
20        45  '2020/12/20'     97       125     243.0
31        60  '2020/12/31'     92       115     243.0
27        60  '2020/12/27'     92       118     241.0
16        60  '2020/12/16'     98       120     215.2
8         30  '2020/12/09'    109       133     195.1
18        45  '2020/12/18'     90       112       NaN
28        60  '2020/12/28'    103       132       NaN

In [32]:
dataset.shape

(32, 5)

In [24]:
dataset = dataset.drop_duplicates(subset=['Pulse', 'Maxpulse'], keep='first')
dataset

Duration          Date  Pulse  Maxpulse  Calories
1         60  '2020/12/02'    117       145     479.0
0         60  '2020/12/01'    110       130     409.1
4         45  '2020/12/05'    117       148     406.0
30        60  '2020/12/30'    102       129     380.3
14        60  '2020/12/14'    104       132     379.3
6         60  '2020/12/07'    110       136     374.0
21        60  '2020/12/21'    108       131     364.2
13        60  '2020/12/13'    106       128     345.3
2         60  '2020/12/03'    103       135     340.0
25        60  '2020/12/25'    102       126     334.5
10        60  '2020/12/11'    103       147     329.3
19        60  '2020/12/19'    103       123     323.0
23        60  '2020/12/23'    130       101     300.0
5         60  '2020/12/06'    102       127     300.0
17        60  '2020/12/17'    100       120     300.0
3         45  '2020/12/04'    109       175     282.4
22        45           NaN    100       119     282.0
29        60  '2020/12/29'    100       132     280.0
15        60  '2020/12/15'     98       123     275.0
9         60  '2020/12/10'     98       124     269.0
7        450  '2020/12/08'    104       134     253.3
24        45  '2020/12/24'    105       132     246.0
20        45  '2020/12/20'     97       125     243.0
31        60  '2020/12/31'     92       115     243.0
27        60  '2020/12/27'     92       118     241.0
16        60  '2020/12/16'     98       120     215.2
8         30  '2020/12/09'    109       133     195.1
18        45  '2020/12/18'     90       112       NaN
28        60  '2020/12/28'    103       132       NaN

In [25]:
dataset.shape

(29, 5)

# Removal of irrelevant data
Sometimes, certain categories/columns in a dataset are not useful. In our case, column name, serial no are not important. Retaining it will take unnecessary space and consume time. 

Panda provides an easy command del to remove the unwanted column.

In [26]:
del dataset['Pulse']

In [28]:
dataset

Duration          Date  Maxpulse  Calories
1         60  '2020/12/02'       145     479.0
0         60  '2020/12/01'       130     409.1
4         45  '2020/12/05'       148     406.0
30        60  '2020/12/30'       129     380.3
14        60  '2020/12/14'       132     379.3
6         60  '2020/12/07'       136     374.0
21        60  '2020/12/21'       131     364.2
13        60  '2020/12/13'       128     345.3
2         60  '2020/12/03'       135     340.0
25        60  '2020/12/25'       126     334.5
10        60  '2020/12/11'       147     329.3
19        60  '2020/12/19'       123     323.0
23        60  '2020/12/23'       101     300.0
5         60  '2020/12/06'       127     300.0
17        60  '2020/12/17'       120     300.0
3         45  '2020/12/04'       175     282.4
22        45           NaN       119     282.0
29        60  '2020/12/29'       132     280.0
15        60  '2020/12/15'       123     275.0
9         60  '2020/12/10'       124     269.0
7        450  '2020/12/08'       134     253.3
24        45  '2020/12/24'       132     246.0
20        45  '2020/12/20'       125     243.0
31        60  '2020/12/31'       115     243.0
27        60  '2020/12/27'       118     241.0
16        60  '2020/12/16'       120     215.2
8         30  '2020/12/09'       133     195.1
18        45  '2020/12/18'       112       NaN
28        60  '2020/12/28'       132       NaN

# Cleaning Strings in Pandas

In [27]:
# Loading a Sample Pandas DataFrame
import pandas as pd
df = pd.DataFrame.from_dict({
    'Name': ['Tranter, Melvyn', 'Lana, Courtney', 'Abel, Shakti', 'Vasu, Imogene', 'Aravind, Shelly'],
    'Region': ['Region A', 'Region A', 'Region B', 'Region C', 'Region D'],
    'Location': ['TORONTO', 'LONDON', 'New york', 'ATLANTA', 'toronto'],
    'Favorite Color': ['   green  ', 'red', '  yellow', 'blue', 'purple  ']
})

In [28]:
df

Name    Region  Location Favorite Color
0  Tranter, Melvyn  Region A   TORONTO        green  
1   Lana, Courtney  Region A    LONDON            red
2     Abel, Shakti  Region B  New york         yellow
3    Vasu, Imogene  Region C   ATLANTA           blue
4  Aravind, Shelly  Region D   toronto       purple

In [29]:
df["Favorite Color"].str.len()

0    10
1     3
2     8
3     4
4     8
Name: Favorite Color, dtype: int64

We can see that our DataFrame has some messy string data! 
For example, some columns contain multiple data points (first and last name), 
others have redundant data (the word ‘Region’), have messy capitalization (location),
and have added whitespace (favorite colors).

In order to apply string methods to an entire Pandas Series, 
you need to use the .str attribute, which gives you access to apply vectorized string methods.

Trimming White Space in Pandas Strings

In [30]:
# Trimming Whitespace from a Pandas Column
df['Favorite Color'] = df['Favorite Color'].str.strip()
df

Name    Region  Location Favorite Color
0  Tranter, Melvyn  Region A   TORONTO          green
1   Lana, Courtney  Region A    LONDON            red
2     Abel, Shakti  Region B  New york         yellow
3    Vasu, Imogene  Region C   ATLANTA           blue
4  Aravind, Shelly  Region D   toronto         purple

In [31]:
#we were able to successfully strip whitespace from a column by re-assigning it to itself.
df["Favorite Color"].str.len()

0    5
1    3
2    6
3    4
4    6
Name: Favorite Color, dtype: int64

# Splitting Strings into Columns in Pandas
The 'Name' column contains both the person’s last and first names. In many cases, you may want to split this column into two – one for each the first and last name. This approach will work a little differently, as we will want to assign two columns, rather than just one.

In [32]:
# Applying .split on a name column
print(df['Name'].str.split(','))

0    [Tranter,  Melvyn]
1     [Lana,  Courtney]
2       [Abel,  Shakti]
3      [Vasu,  Imogene]
4    [Aravind,  Shelly]
Name: Name, dtype: object


In [33]:
# Splitting a Column into Two Columns
df[['Last Name', 'First Name']] = df['Name'].str.split(',', expand=True)

print(df)

              Name    Region  Location Favorite Color Last Name First Name
0  Tranter, Melvyn  Region A   TORONTO          green   Tranter     Melvyn
1   Lana, Courtney  Region A    LONDON            red      Lana   Courtney
2     Abel, Shakti  Region B  New york         yellow      Abel     Shakti
3    Vasu, Imogene  Region C   ATLANTA           blue      Vasu    Imogene
4  Aravind, Shelly  Region D   toronto         purple   Aravind     Shelly


In [34]:
# Replacing a Substring in Pandas
df['Region'] = df['Region'].str.replace('Region ', '')
print(df)

              Name Region  Location Favorite Color Last Name First Name
0  Tranter, Melvyn      A   TORONTO          green   Tranter     Melvyn
1   Lana, Courtney      A    LONDON            red      Lana   Courtney
2     Abel, Shakti      B  New york         yellow      Abel     Shakti
3    Vasu, Imogene      C   ATLANTA           blue      Vasu    Imogene
4  Aravind, Shelly      D   toronto         purple   Aravind     Shelly


# Exercise

# 1.  Calculate the percentage of missing records in each column

# 2. Drop any duplicate record based on the name column, keeping the last record

# Create a fisrt name & a last name column. Note that there is semicolon between names